File: src/org/cerion/symcalc/expression/NumberExpr.java
package org.cerion.symcalc.expression;

import org.cerion.symcalc.expression.number.ComplexNum;
import org.cerion.symcalc.expression.number.IntegerNum;
import org.cerion.symcalc.expression.number.RealNum;

public abstract class NumberExpr extends Expr
{
	//Types
	public static final int INTEGER = 0;
	public static final int RATIONAL = 1;
	public static final int REAL = 2;
	public static final int COMPLEX = 3;
	public abstract int numType(); //getType already used by MathTerm
	
	public abstract String toString();
	public abstract NumberExpr add(NumberExpr num);
	public abstract NumberExpr subtract(NumberExpr num);
	public abstract NumberExpr multiply(NumberExpr num);
	public abstract NumberExpr divide(NumberExpr num);
	public abstract NumberExpr power(NumberExpr num);
	public abstract boolean canExp(NumberExpr num); //this^num = num is TRUE, FALSE if can't resolve
	public abstract NumberExpr negate();

	public abstract double toDouble(); //Valid on all but ComplexNum
	public abstract boolean isZero();
	public abstract boolean isOne();
	public abstract boolean equals(NumberExpr e);

	//Members
	//private BigInteger n;

	@Override
	public boolean equals(Expr e) {
		return e.isNumber() && equals((NumberExpr)e);
	}

	@Override
	public boolean isInteger() {
		return numType() == INTEGER;
	}

	public boolean isReal() {
		return numType() == REAL;
	}

	public boolean isRational() {
		return numType() == RATIONAL;
	}

	public boolean isComplex() {
		return numType() == COMPLEX;
	}

	//Inherited from MathTerm
	public ExprType getType() { return ExprType.NUMBER; }
	public void show(int i) 
	{
		indent(i,"Number " + this.toString());
	}

	@Override
	public NumberExpr evaluate() {
		return this;
	}

	public static NumberExpr parse(String s) {
		if(s.indexOf('i') > -1)
			return new ComplexNum(s);

		if(s.indexOf('.') > 0)
			return RealNum.parse(s);

		return new IntegerNum(s);
	}
}

File: src/org/cerion/symcalc/expression/NumberExpr.kt
package org.cerion.symcalc.expression

import org.cerion.symcalc.expression.number.ComplexNum
import org.cerion.symcalc.expression.number.IntegerNum
import org.cerion.symcalc.expression.number.RationalNum
import org.cerion.symcalc.expression.number.RealNum

abstract class NumberExpr : Expr(), Comparable<NumberExpr> {

    override val type: ExprType
        get() = ExprType.NUMBER

    override val isInteger: Boolean
        get() = numType() == INTEGER

    abstract val isZero: Boolean
    abstract val isOne: Boolean

    val isReal: Boolean
        get() = numType() == REAL

    val isRational: Boolean
        get() = numType() == RATIONAL

    val isComplex: Boolean
        get() = numType() == COMPLEX

    fun asRational(): RationalNum = this as RationalNum
    fun asComplex(): ComplexNum = this as ComplexNum

    abstract fun numType(): Int  //getType already used by MathTerm

    abstract override fun toString(): String
    abstract fun power(num: NumberExpr): Expr
    abstract fun canExp(num: NumberExpr): Boolean  //this^num = num is TRUE, FALSE if can't resolve
    abstract fun toDouble(): Double  //Valid on all but ComplexNum
    abstract fun equals(e: NumberExpr): Boolean

    override fun equals(e: Expr): Boolean = e.isNumber && equals(e as NumberExpr)
    override fun show(i: Int) = indent(i, "Number " + this.toString())
    public override fun evaluate(): NumberExpr = this

    abstract operator fun plus(number: NumberExpr): NumberExpr
    abstract operator fun minus(number: NumberExpr): NumberExpr
    abstract operator fun times(num: NumberExpr): NumberExpr
    abstract operator fun div(num: NumberExpr): NumberExpr
    abstract operator fun unaryMinus(): NumberExpr

    companion object {
        //Types
        // TODO make this enum
        const val INTEGER = 0
        const val RATIONAL = 1
        const val REAL = 2
        const val COMPLEX = 3

        @JvmStatic fun parse(s: String): NumberExpr {
            if (s.indexOf('i') > -1)
                return ComplexNum(s)

            return if (s.indexOf('.') > 0) RealNum.parse2(s) else IntegerNum(s)

        }
    }
}
